src/engine/client/key_identification.cpp

```diff
@@ -0,0 +1,504 @@
+{
+    if (!Q_stricmp(s.data(), "SEMICOLON"))
+        return ';';
+    if (!Q_stricmp(s.data(), "BACKSLASH"))
+        return '\\';
+    if (!Q_stricmp(s.data(), "SPACE"))
+        return ' ';
+    if (Q_UTF8_Strlen(s.data()) == 1)
+        return Str::ctolower(Str::UTF8To32(s.data())[0]);
+    return 0;
+}
+
+// This takes a character name and tries to find a key that types that character.
+// If the SDL lookup function doesn't find one, fall back to the scancode
+// for the key on a QWERTY layout, if possible.
+static Key KeyFromUnprefixedCharacter(int ch)
+{
+    SDL_Scancode sc = SDL_GetScancodeFromKey(Util::enum_cast<SDL_Keycode>(ch));
+    if (sc != SDL_SCANCODE_UNKNOWN) {
+        return Key::FromScancode(sc);
+    }
+    return Key::FromScancode(AsciiToScancode(ch));
+}
+
+/*
+===================
+Key_StringToKeynum
+
+Returns a key number to be used to index keys[] by looking at
+the given string.  Single ascii characters return themselves, while
+the K_* names are matched up.
+
+0x11 will be interpreted as raw hex, which will allow new controlers
+
+to be configured even if they don't have defined names.
+===================
+*/
+Key Key_StringToKeynum( const char *str )
+{
+	const keyname_t *kn;
+
+	if ( !str )
+	{
+		return Key::NONE;
+	}
+
+	if ( int c = ParseCharacter( str ) )
+	{
+		return KeyFromUnprefixedCharacter( c );
+	}
+
+	// check for hex code
+	if ( str[0] == '0' && str[1] == 'x' )
+	{
+		int n = Com_HexStrToInt( str );
+		return Key::FromScancode( n );
+	}
+
+    // Physical key by QWERTY location, from ascii char
+    int prefixLen = strlen(SCANCODE_ASCII_BIND_PREFIX);
+    if ( !Q_strnicmp( SCANCODE_ASCII_BIND_PREFIX, str, prefixLen ) && strlen( str ) == prefixLen + 1 )
+    {
+        return Key::FromScancode( AsciiToScancode( ParseCharacter ( str + prefixLen ) ) );
+    }
+
+    // char:X forces a virtual key-based rather than scancode binding to be used for some (Unicode) char
+    prefixLen = strlen( CHARACTER_BIND_PREFIX );
+    if ( !Q_strnicmp( CHARACTER_BIND_PREFIX, str, prefixLen ) && Q_UTF8_Strlen( str + prefixLen ) == 1 )
+    {
+        return Key::FromCharacter( ParseCharacter( str + prefixLen ) );
+    }
+
+	// scan for a text match
+	for ( kn = keynames; kn->name; kn++ )
+	{
+		if ( !Q_stricmp( str, kn->name ) )
+		{
+			return kn->keynum;
+		}
+	}
+
+	return Key::NONE;
+}
+
+
+
+
+/*
+===================
+Key_KeynumToString
+
+Returns a string (either a single ascii char, a K_* name, or a 0x11 hex string) for the
+given keynum.
+===================
+*/
+const char *Key_KeynumToString( Key key )
+{
+	const keyname_t *kn;
+	static std::string tinystr;
+
+    // check for a key string
+	for ( kn = keynames; kn->name; kn++ )
+	{
+		if ( key == kn->keynum )
+		{
+			return kn->name;
+		}
+	}
+
+	if ( key.kind() == Key::Kind::UNICODE_CHAR )
+	{
+		tinystr = CHARACTER_BIND_PREFIX;
+        tinystr += Q_UTF8_Encode( key.AsCharacter() );
+        tinystr = Cmd::Escape( tinystr );
+        return tinystr.data();
```

Comment:
Any reason you use data() instead of c_str()? Since it is being used as a string, using c_str() makes the intention more clear.